src/main/kotlin/org/jetbrains/research/transformations/anonimization/Anonymization.kt

```diff
@@ -0,0 +1,164 @@
+
+    // Check if the node has the parent function and get its name
+    // Otherwise return EMPTY_PREFIX
+    private fun getNodeLabelPrefix(node: ITree, oldPrefix: String,
+                                   metaInformation: AnonymizationMetaInformation): String {
+        var prefix = EMPTY_PREFIX
+        val parentFunctions = node.parents.map { it.label }.filter { it.isNotEmpty() }
+        if (parentFunctions.isNotEmpty()) {
+            prefix = parentFunctions[0] + SEPARATOR
+        }
+        if (oldPrefix != prefix) {
+            // Reset ids
+            metaInformation.anonVariablesMeta.currentId = 0
+            metaInformation.anonArgumentsMeta.currentId = 0
+        }
+        return prefix
+    }
+
+    private fun anonymizeVariable(node: ITree,
+                                  anonymousConfig: AnonymousConfig,
+                                  currentPrefix: String,
+                                  toStoreMetadata: Boolean) {
+        anonymousConfig.namesMap[node.label]?.forEach { (_, parentPrefix) ->
+            if (parentPrefix == currentPrefix) {
+                setAnonLabel(node, currentPrefix, toStoreMetadata)
+            }
+        }
+    }
+
+    private fun initOrFindLabel(node: ITree,
+                                anonymousConfig: AnonymousConfig,
+                                currentPrefix: String,
+                                toStoreMetadata: Boolean) {
+        val prefixes = anonymousConfig.getPrefixes(node)
+        if ((node.label in anonymousConfig.namesMap.keys) and (currentPrefix in prefixes)) {
+            anonymizeVariable(node, anonymousConfig, currentPrefix, toStoreMetadata)
+        } else {
+            val newLabel = setAnonLabel(node, currentPrefix, toStoreMetadata,
+                    getMarkedLabel(NodeType.NAME_STORE, anonymousConfig.currentId))
+            anonymousConfig.initLabel(node.label, newLabel, currentPrefix)
+        }
+    }
+
+    private fun handleNameLoadNode(node: ITree,
+                                   metaInformation: AnonymizationMetaInformation,
+                                   currentPrefix: String,
+                                   toStoreMetadata: Boolean) {
+        when (node.label) {
+            in metaInformation.anonVariablesMeta.namesMap.keys -> {
+                anonymizeVariable(node, metaInformation.anonVariablesMeta, currentPrefix, toStoreMetadata)
+            }
+            in metaInformation.anonFunctionsMeta.namesMap.keys -> {
+                val newLabel = metaInformation.anonFunctionsMeta.namesMap[node.label]
+                newLabel?.let {
+                    setAnonLabel(node, newLabel, toStoreMetadata)
+                }
+            }
+            in metaInformation.anonArgumentsMeta.namesMap.keys -> {
+                anonymizeVariable(node, metaInformation.anonArgumentsMeta, currentPrefix, toStoreMetadata)
+            }
+        }
+    }
+
+    private fun handleFuncDefinitionNode(node: ITree,
```

Comment:
Shouldn't we also keep the first underscores if they were in the name to keep the information about private methods?

Reply:
I am not sure, but we can